WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java

```diff
@@ -1221,27 +1207,36 @@ private void savePostAndFinish() {
                 }
             }
 
-            if (PostStatus.fromPost(mPost) != PostStatus.PUBLISHED && isPublishable) {
-                publishPost();
+            if (PostStatus.fromPost(mPost) != PostStatus.PUBLISHED && isPublishable && NetworkUtils.isNetworkAvailable(this)) {
+                if (!hasFailedMedia()) {
+                    savePostOnlineAsyncAndFinish(isFirstTimePublish);
+                }
             } else {
-                savePostLocally();
+                savePostLocallyAsyncFinish();
             }
+        } else {
+            // discard post if new & empty
+            if (!isPublishable && isNewPost()) {
+                mDispatcher.dispatch(PostActionBuilder.newRemovePostAction(mPost));
+            }
+            finish();
         }
+    }
 
-        // discard post if new & empty
-        if (isNewPost() && !isPublishable) {
-            mDispatcher.dispatch(PostActionBuilder.newRemovePostAction(mPost));
-        }
+    private boolean updatePostObject() {
```

Comment:
> Because of these, I would suggest splitting this into a boolean isFirstTimePublish() and a void updatePostObject() for clearer functionality.

Good point.

> Also, it seems that it catches the IllegalEditorStateException, thus allowing its caller to continue normally, without a user notification. Not sure at this point but, can you confirm that this is indeed the intended behaviour here? The other version of updatePostObject does throw it. Thanks!

You're right. It doesn't make sense to continue if the post update fails. I fixed the flow to account for this.